src/test/pmemset_source/pmemset_source.c

```diff
@@ -436,6 +436,246 @@ test_src_from_file_with_truncate(const struct test_case *tc, int argc,
+	struct pmemset_source *src;
+	uint64_t flags = 0;
+	int ret = 0;
+
+	flags = PMEMSET_SOURCE_FILE_CREATE_ALWAYS | \
+		PMEMSET_SOURCE_FILE_CREATE_MODE(90180);
+	ret = pmemset_xsource_from_file(&src, file, flags);
+	UT_PMEMSET_EXPECT_RETURN(ret,
+		PMEMSET_E_INVALID_SOURCE_FILE_CREATE_FLAGS);
+	UT_ASSERTeq(src, NULL);
+
+	ret = pmemset_source_delete(&src);
+	UT_PMEMSET_EXPECT_RETURN(ret, 0);
+	UT_ASSERTeq(src, NULL);
+
+	return 1;
+}
+
+/*
+ * test_src_from_file_only_mode - test source creation with
+ * only mode value in flag param.
+ */
+static int
+test_src_from_file_only_mode(const struct test_case *tc, int argc,
+		char *argv[])
+{
+	if (argc < 1)
+		UT_FATAL("usage: test_src_from_file_only_mode <path>");
+
+	const char *file = argv[0];
+	struct pmemset_source *src;
+	uint64_t flags = 0;
+	int ret = 0;
+/* just to compile test on Windows */
+#ifndef S_IXUSR
+#define S_IXUSR 00100
+#endif
+	flags = PMEMSET_SOURCE_FILE_CREATE_ALWAYS_MODE(S_IXUSR);
+	ret = pmemset_xsource_from_file(&src, file, flags);
+	UT_PMEMSET_EXPECT_RETURN(ret, 0);
+	UT_ASSERTne(src, NULL);
+
+	ret = os_access(file, F_OK | R_OK | X_OK);
+	UT_ASSERTeq(ret, -1);
+
+	ret = os_access(file, F_OK | X_OK);
+	UT_ASSERTeq(ret, 0);
+
+	ret = pmemset_source_delete(&src);
+	UT_PMEMSET_EXPECT_RETURN(ret, 0);
+	UT_ASSERTeq(src, NULL);
+
+	return 1;
+}
+
+/*
+ * test_src_from_file_with_inval_win_mode - test source creation with
+ * invalid mode value on Windows in PMEMSET_SOURCE_FILE_CREATE_ALWAYS_MODE.
+ */
+static int
+test_src_from_file_with_inval_win_mode(const struct test_case *tc, int argc,
+	char *argv[])
+{
+	if (argc < 1)
+		UT_FATAL(
+			"usage: test_src_from_file_with_inval_win_mode <path>");
+
+	const char *file = argv[0];
+	struct pmemset_source *src;
+	uint64_t flags = 0;
+	int ret = 0;
+
+	/* "random" flag does not work */
+	flags = PMEMSET_SOURCE_FILE_CREATE_ALWAYS_MODE(00100);
+	ret = pmemset_xsource_from_file(&src, file, flags);
+	UT_PMEMSET_EXPECT_RETURN(ret,
+		PMEMSET_E_INVALID_SOURCE_FILE_CREATE_FLAGS);
+	UT_ASSERTeq(src, NULL);
+
+	/* PMEMSET flag works but do nothing internally */
+	flags = PMEMSET_SOURCE_FILE_CREATE_ALWAYS_MODE(
+			PMEMSET_SOURCE_FILE_RWXU_MODE);
+	ret = pmemset_xsource_from_file(&src, file, flags);
+	UT_PMEMSET_EXPECT_RETURN(ret, 0);
+	UT_ASSERTne(src, NULL);
+
+	ret = pmemset_source_delete(&src);
+	UT_PMEMSET_EXPECT_RETURN(ret, 0);
+	UT_ASSERTeq(src, NULL);
+
+	return 1;
+}
+
+/*
+ * test_src_from_file_with_rusr_mode_if_needed - test source creation with
+ * PMEMSET_SOURCE_FILE_CREATE_ALWAYS_MODE flags.
+ */
+static int
+test_src_from_file_with_rusr_mode_if_needed(const struct test_case *tc,
+		int argc, char *argv[])
+{
+	if (argc < 1)
+		UT_FATAL(
+			"usage: test_src_from_file_with_rusr_mode_if_needed <path>");
+
+	const char *file = argv[0];
+	struct pmemset_source *src;
+	uint64_t flags = 0;
+	int ret = 0;
+
+	flags = PMEMSET_SOURCE_FILE_CREATE_IF_NEEDED | \
+	PMEMSET_SOURCE_FILE_CREATE_MODE(PMEMSET_SOURCE_FILE_RUSR_MODE);
```

Comment:
I think, It might be a good idea to add a test case when both these flags are used but the file already exists.